lib/Backends/BackendUtils.cpp

```diff
@@ -81,6 +74,55 @@ runtime::RuntimeBundle::getSymbolInfo(const Named *v) const {
   return it->second;
 }
 
+/// If \p PH is an output placeholder, \returns true.
+/// This is determined by checking if the PH has a user which uses the PH as an
+/// overwritten input.
+bool isOutput(const Placeholder *PH) {
+  // Check if trainable, if it is return true.
+  if (PH->isTraining()) {
+    return true;
+  }
+  for (auto &use : PH->getUsers()) {
+    // Look through the inputs of the PH's users. If an input is overwritten
+    // check if it's the PH, if it is return true.
+    for (unsigned i = 0; i < use.getUser()->getNumInputs(); i++) {
+      // If the input is not overwritten we can continue.
+      if (!use.getUser()->isOverwrittenNthInput(i)) {
+        continue;
+      }
+      auto input = use.getUser()->getNthInput(i);
+      if (input == PH) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+/// If \p PH is an input placeholder, \returns true.
+bool isInput(const Placeholder *PH) {
+  // Check if trainable, if it is return true.
+  if (PH->isTraining()) {
+    return true;
+  }
+  for (auto &use : PH->getUsers()) {
+    // Check if PH is a non overwritten input to a saveNode.
+    if (auto save = llvm::dyn_cast<SaveNode>(use.getUser())) {
```

Comment:
You can add 

using llvm::dyn_cast; 

at the end of the file so that you won't need to type the llvm namespace every time we use cast or dyn_cast. We do this in a few places in the compiler.